src/outputs/rcp.c

```diff
@@ -0,0 +1,1464 @@
+}
+
+
+static void
+rcp_session_free(struct rcp_session *s)
+{
+  if (!s)
+    return;
+
+  DPRINTF(E_DBG, L_RCP, "session free %ld %s at %s\n", s->device->id, s->devname, s->address);
+
+
+  if (s->sock >= 0)
+    rcp_disconnect(s->sock);
+  if (s->ev)
+    event_free(s->ev);
+  if (s->reply_timeout)
+    event_free(s->reply_timeout);
+
+  free(s->stream_url);
+  free(s->address);
+  free(s->devname);
+
+  free(s);
+}
+
+static void
+rcp_session_cleanup(struct rcp_session *rs)
+{
+  struct rcp_session *s;
+
+  DPRINTF(E_DBG, L_RCP, "session cleanup %ld %s at %s\n", rs->device->id, rs->devname, rs->address);
+
+  if (rs == rcp_sessions)
+    rcp_sessions = rcp_sessions->next;
+  else
+    {
+      for (s = rcp_sessions; s && (s->next != rs); s = s->next)
+	; /* EMPTY */
+
+      if (!s)
+	DPRINTF(E_WARN, L_RCP, "WARNING: struct rcp_session not found in list; BUG!\n");
+      else
+	s->next = rs->next;
+    }
+
+  outputs_device_session_remove(rs->device->id);
+
+  rcp_session_free(rs);
+}
+
+
+/* ---------------------------- STATUS HANDLERS ----------------------------- */
+
+static void
+rcp_status(struct rcp_session *s)
+{
+  enum output_device_state state;
+
+  switch (s->state)
+    {
+      case RCP_STATE_SETUP:
+      case RCP_STATE_SETUP_WAKEUP:
+	state = OUTPUT_STATE_STARTUP;
+	break;
+
+      case RCP_STATE_SETUP_GET_CONNECTED_SERVER ... RCP_STATE_SETUP_VOL_GET:
+      case RCP_STATE_QUEUING_CLEAR ... RCP_STATE_QUEUING_PLAY:
+      case RCP_STATE_VOL_GET:
+      case RCP_STATE_VOL_SET:
+      case RCP_STATE_STOPPING ... RCP_STATE_STANDBY:  // as per @frankusb, move advise that remote/we're not interested
+	state = OUTPUT_STATE_CONNECTED;
+	break;
+
+      case RCP_STATE_STREAMING:
+	state = OUTPUT_STATE_STREAMING;
+	break;
+
+      case RCP_STATE_DISCONNECTED:
+	state = OUTPUT_STATE_STOPPED;
+	break;
+
+      case RCP_STATE_FAILED:
+      default:
+	state = OUTPUT_STATE_FAILED;
+    }
+
+  DPRINTF(E_DBG, L_RCP, "mapping state from (internal) %d -> (output) %d\n", s->state, state);
+  outputs_cb(s->callback_id, s->device->id, state);
+  s->callback_id = -1;
+
+  if (state == OUTPUT_STATE_STOPPED || state == OUTPUT_STATE_FAILED)
+    rcp_session_cleanup(s);
+}
+
+
+/* ------------------ INTERFACE FUNCTIONS CALLED BY OUTPUTS.C --------------- */
+
+static int
+rcp_device_start(struct output_device *device, int callback_id)
+{
+  struct rcp_session *s;
+
+  DPRINTF(E_DBG, L_RCP, "device start %ld '%s'\n", device->id, device->name);
+
+  s = rcp_session_make(device, callback_id);
+  if (!s)
+    return -1;
+
+  return 1;
+}
+
+static int
+rcp_device_stop(struct output_device *device, int callback_id)
+{
+  struct rcp_session *s = device->session;
+
+  DPRINTF(E_DBG, L_RCP, "device stop %ld '%s'\n", device->id, device->name);
+
+  /* force these devices as deselected (auto state saves in db later) since 
+   * these need use to select (and cause the device probe to start connection to
+   * remote side
+   */
+  device->selected = 0;
```

Comment:
Normally, changing selected status is only something the user is allowed to do. I didn't understand the comment above, so I suppose it's possible that there is justification for it here. Can you check if you can use other flags instead? For instance "prevent_playback" or "busy". If not, try to make the explanation more clear.

Reply:
using `prevent_playback` also works